### ISSUE TITLE
Support zlib compression for TGA, save 95% on transfer size

### ISSUE THREAD

**Issue by erlehmann** [minetest/minetest#13795]
### Problem

I want generated textures to be a smaller in transit
===

The 512×512 PNG checkerboard generated by devtest is over 5000 bytes. Large for a simple checkerboard!

The texture files generated by `tga_encoder` (maps, signs) could also be a bit smaller, if you ask me.

### Solutions

How to make generated textures small (~5% size, slow)
===

**`optipng -o7` reduces the checkerboard to about 240 bytes.**

But `optipng` takes over half a second for it on my machine. Very slow!

Also I  think I can not call optipng from a MInetest mod that generates textures.

(If you have read until here you may have noticed that this is not a suitable solution.)

How to make generated textures small (~4% size, fast)
===

**As a 32bpp (RGBA) RLE TGA compressed with zlib the checkerboard is only about 200 bytes.**

Compressing using zlib is almost instantaneously, beats optipng on size and speed.

I think such a compression is available from `minetest.compress()`.

A file can be created, Minetest does not understand.

Implementation proposal
===

Patch the client so that files ending with `.TGA.Z`  (case-insensitive) are decoded using zlib and then by regular TGA code.

The server should be able to send a zlib-compressed TGA to a new client and an uncompressed TGA file to an old client that does not support this method.

Don't automatically compress files.

### Alternatives

* Make the PNG encoder use optipng, wait half a second each for your optimized textures.
* Do nothing. Enjoy your really big (like 25 times the size they should be) textures.

### Additional context

Saving 95% of transfer size is a huge thing!
===

Also isn't it funny that you can beat `optipng` in both size and speed with *one simple trick*?

Also, I think I can optimize that TGA.Z. checkerboard a bit further, but does it really matter?

**Comment by sfan5** [comment]
Automatically compressing media in transit can be discussed but I do not support tacking a custom modification onto a niche format :-1:

**Comment by erlehmann** [comment]
> Automatically compressing media in transit sounds useful

**Ah yes, ”we should totally do the one thing that an expert said is a bad idea”, again!**

Doing it automatically (most likely) only *sounds* useful. Consider the following:

* A common assumption is that every file would benefit from that. Does it?
  * I could be easily be convinced by benchmarks – but do you have those?
  * Which content or encoding characteristics influence if textures compress well?
    * Could `mcl_maps` or `xmaps` have diffirent requirements than `unicode_text`?
    * How would a pathological case for automatic texture compression look like?
* Ideally textures are only compressed once, not every time they are sent out.
  * To handle this, the automatic approach needs to have a good caching strategy.
  * Note that any cache of compressed files possibly needs to be evicted at runtime.
* Recompressing arbitrary TGAs optimally requires the following:
  * Better TGA introspection than IrrlichtMT already has.
    * This might be needed to decide if you want to compress a TGA at all.
  * A TGA writer on the server side that is at least as good as `tga_encoder`.
* Recompressing at runtime is most likely not the *fastest* way to do it.
  * With a good cache that could only matter for the first texture sent.
    * Unfortunately, the first texture sent is most important one for generated textures.

> but I do not support tacking a custom modification

Compressing files that way and adding an extension to indicate it is a very common thing.

A typical Debian GNU/Linux installation has tens of thousands of `.gz` files in its `/usr` directory.

> onto a niche format 👎.

TGA is indeed quite a niche format – and AFAIK one niche is “textures for 3D models & engines”.

Maybe you have heard of one of the following games/engines that used TGA as a texture format?

* Half Life
* Minecraft
* World of Warcraft
* Godot Engine
* Homeworld 2
* Doom 3
* Quake 4
* Unreal Engine 5

TL;DR:
===
It is likely inefficient, more code, and wastes CPU cycles to recompress textures automatically.

If you think otherwise, you probably have not thought too hard about how to implement it.

**Comment by sfan5** [comment]
> If you think otherwise, you probably have not thought too hard about how to implement it.

Kindly fuck off with your condescending smartass attitude and never talk to me again.
